### ISSUE TITLE
DeepSeek V4 HyperConnection residual mixing should transpose comb

### ISSUE THREAD

**Issue by jQizhang** [NVIDIA-NeMo/Automodel#2154]
**Describe the bug**

In `DeepseekV4Block.forward`, the HyperConnection residual mixing appears to use the HC stream mixing matrix `comb` in the wrong orientation.

The current implementation applies `comb` directly:

```python
torch.matmul(comb.to(dtype), x)
```

at both the attention HC site and the MLP HC site. Based on parity testing against the DeepSeek-V4 native BF16 implementation, this should instead use the transposed `comb`:

```python
torch.matmul(comb.transpose(-1, -2).to(dtype), x)
```

Without the transpose, Automodel forward results can diverge significantly from the native/reference implementation. On a full DeepSeek-V4-Flash fixed-token forward test, this shows up as large logprob outliers and very high generation KL.

**Steps/Code to reproduce bug**

Run a fixed-token logprob comparison between Automodel DeepSeekV4 and a native BF16 DeepSeek-V4 implementation using the same model weights and input tokens.

Observed results vs native BF16:

| Variant | KL(Native || Automodel) |
|---|---:|
| baseline | 0.265243 |
| transpose fix | 0.038195 |

**Comment by jQizhang** [comment]
Draft a PR #2159  for reference. Feel free to leave any comments or feedback.

**Comment by HuiyingLi** [comment]
@jQizhang Thank you so much for capturing this. This is a real bug introduced into our implementation thru porting transformers's implementation. We were referring to an open PR 45616 on transformers back then. Now DSV4 is merged, seems there is still this: https://github.com/huggingface/transformers/blob/da6c53e431f7c9ef0691239d4ce89b0f711ecad7/src/transformers/models/deepseek_v4/modular_deepseek_v4.py#L948
We changed many things while porting transformers but seem to have missed it, as the cosine sim delta it introduced in our logits comparison is very small. I really appreciate your PR. Thank you!
